tinymce-basics/badges.html

```diff
@@ -0,0 +1,245 @@
+                            node = editor.selection.getNode();
+
+                            if (editor.dom.hasClass(node, 'badge')) {
+                                return node.style.color;
+                            }
+
+                            return 'rgb(255, 255, 255)';
+                        }
+
+                        // Open a TinMCE modal where the user can set the badge's
+                        // background and text color.
+                        // https://www.tiny.cloud/docs/ui-components/dialog/
+                        // https://www.tiny.cloud/docs/ui-components/dialogcomponents/
+                        editor.windowManager.open({
+                            title: 'Insert/edit Badge',
+                            body: {
+                                type: 'panel',
+                                items: [
+                                    {
+                                        type: 'colorinput', // component type
+                                        name: 'backgroundcolor', // identifier
+                                        label: 'Background color' // text for the label
+                                    },
+                                    {
+                                        type: 'colorinput', // component type
+                                        name: 'textcolor', // identifier
+                                        label: 'Text color' // text for the label
+                                    }
+                                ]
+                            },
+                            buttons: [
+                            {
+                                type: 'cancel',
+                                name: 'closeButton',
+                                text: 'Cancel'
+                            },
+                            {
+                                type: 'submit',
+                                name: 'submitButton',
+                                text: 'Save',
+                                primary: true
+                            }
+                            ],
+                            initialData: {
+                                backgroundcolor: backgroundColor(),
+                                textcolor: textColor()
+                            },
+                            onSubmit: function (dialog) {
+                                // Get the form data.
+                                var data = dialog.getData();
+
+                                // Get the currently selected node.
+                                // https://www.tiny.cloud/docs/api/tinymce.dom/tinymce.dom.selection/
+                                const node = editor.selection.getNode();
+
+                                // Check if the selected node is a badge.
+                                // https://www.tiny.cloud/docs/api/tinymce.dom/tinymce.dom.domutils/
+                                if (editor.dom.hasClass(node, 'badge')) {
```

Comment:
This should probably be wrapped in an undo transaction otherwise users won't be able to undo changes.

Reply:
@lnewson Something like this?
```
// Working directly with the API often requires manually adding
// the actions to the undo stack.
// https://www.tiny.cloud/docs/api/tinymce/tinymce.undomanager/
editor.undoManager.transact(() => {
    ...
});
```

 Its vague because I don't really know how it works and when it's needed. And the docs isn't super informative either...

Reply:
Yeah, that's all that's needed. The general rule is that if you're editing the DOM directly in any form, then you need to wrap it in a transaction.